Ingest ES structured audit logs

ycombinator opened the pull request:
This is a "forward port" of https://github.com/elastic/beats/pull/8852.

In https://github.com/elastic/beats/pull/8852, we taught Filebeat to ingest either structured or unstructured ES audit logs but the resulting fields conformed to the 6.x mapping structure.

In this PR we also teach Filebeat to ingest either structured or unstructured ES audit logs but the resulting fields conform to the 7.0 (ECS-based) mapping structure.

ycombinator commented:
While working on this PR I realized that we don't have sample log lines containing a request body (which is supposed to be parsed into the `http.request.body.content` field). I'm working with the ES Security team to get such a sample and will incorporate it into follow up PRs (for `master` and `6.x`).

ycombinator commented:
jenkins, test this

ruflin commented on filebeat/module/elasticsearch/audit/_meta/fields.yml:
```diff
@@ -6,15 +6,15 @@
       example": "active_directory"
       type: keyword
-    - name: roles
+    - name: user.roles
```
Is the idea with these two two be ecs compatible? If yes, these are not on the root level but nested under elasticsearch.audit.

ycombinator replied:
No, these are meant to be under `elasticsearch.audit`. I looked at https://github.com/elastic/ecs and didn't find a `user.roles` in there.

webmat replied:
Yeah this looks like they are in the right place, under es.audit.*

ycombinator commented:
> Can you also update the ecs-migration.yml file?

Updated now. Do I need to run any `make`/`mage` commands as well?

ycombinator commented:
@ruflin I'm not sure what you meant by https://github.com/elastic/beats/pull/10352#pullrequestreview-196902914:

> Can you also update the ecs-migration.yml file?

Could you please clarify? Otherwise this PR is ready for review, IMHO.

ycombinator commented:
jenkins, test this

webmat commented on filebeat/module/elasticsearch/audit/ingest/pipeline.json:
```diff
@@ -10,56 +10,29 @@
         {
             "date": {
-                "field": "elasticsearch.audit.timestamp",
+                "field": "elasticsearch.audit.@timestamp",
```
Nit: prior to grabbing the real timestamp from the log, could you populate `event.created` with Beat's `@timestamp`?

ycombinator replied:
This is already being done in the very first processor in this pipeline. It's collapsed in the diff since nothing changed there:

https://github.com/elastic/beats/pull/10352/files#diff-a007340015953dcf31f9d48f74358ac3R4

webmat replied:
Missed that 👍

ycombinator commented on filebeat/module/elasticsearch/audit/ingest/pipeline-json.json:
```diff
@@ -0,0 +1,131 @@
+        {
+            "rename": {
+                "field": "elasticsearch.audit.user.name",
+                "target_field": "user.name"
```
I'm not sure I follow what you mean by doing "dot_expand in place"? The `dot_expander` processor right above this one is necessary to go from:

```
{
  "elasticsearch.audit.user.name": "foo"
}
```

to:

```
{
  "elasticsearch.audit.user": {
    "name": "foo"
  }
}
```

That then allows us to call the `rename` processor as we are doing over here.

webmat replied:
I haven't used `dot_expander` yet, so perhaps I'm misunderstanding it. But I was under the impression that the following did the equivalent of the 2 processors above:

```JSON
{ "dot_expander": { "field": "user.name" } }
```

And in cases where the output isn't the object equivalent of the dotted notation, you would use `path` this way, to get the equivalent of the two `node.name` processors above:

```JSON
{ "dot_expander": { "field": "node.name", "path": "elasticsearch.node" } }
```

If that's not the case, you can ignore this ;-)

ycombinator commented:
@webmat Thanks for the detailed explanation of how to use the `ecs-migration.yml` file — really appreciate it written down so I can go back and look at it in the future as well!

I've addressed all your comments in the review. Only one of them resulted in a code change, however. So you might want to look at my replies to your other two comments. 

Thanks!
